File: templates/nodejs-service/service/src/core/mainApp.ts
import { Express } from "express";
import { App, Comp } from "./interfaces";
import { logExpressRoutes } from "./logExpressRoutes";
import { Subscriber, RPCServer } from "../transport";

export interface MainAppConfig {
    logger: Console,
    router: Express,
    subscribers: Subscriber[],
    rpcServers: RPCServer[],
    httpPort: number,
}

export class MainApp implements App {

    private _readiness: boolean;
    private _liveness: boolean;
    private _httpPort: number;
    private _router: Express;
    private _rpcServers: RPCServer[];
    private _subscribers: Subscriber[];
    private _logger: Console;


    constructor(config: MainAppConfig) {
        this._httpPort = config.httpPort;
        this._readiness = false;
        this._liveness = false;
        this._logger = config.logger;
        this._router = config.router;
        this._subscribers = config.subscribers;
        this._rpcServers = config.rpcServers;
    }

    setup(components: Comp[]) {
        for (let component of components) {
            component.setup();
        }
    }

    run() {
        const pRouter = new Promise((resolve, reject) => {
            this._router.listen(this._httpPort, () => {
                this._logger.info(`Run at port ${this._httpPort}`);
                logExpressRoutes(this._router, this._logger);
                this._liveness = true;
                this._readiness = true;
                resolve();
            }).on("error", (err) => reject(err));
        });
        const promises: Promise<any>[] = [pRouter];
        for (let rpcServer of this._rpcServers) {
            promises.push(rpcServer.serve());
        }
        for (let subscriber of this._subscribers) {
            promises.push(subscriber.subscribe())
        }
        Promise.all(promises).catch((err) => {
            this._liveness = false;
            this._readiness = false;
            this._logger.error(err);
        });
    }

    liveness() {
        return this._liveness;
    }

    setLiveness(liveness: boolean) {
        this._liveness = liveness;
    }

    readiness() {
        return this._readiness;
    }

    setReadiness(readiness: boolean) {
        this._readiness = readiness;
    }

}

File: templates/python-service/service/core/mainApp.py
from .interfaces import App, Comp
import threading, pika
from flask import Flask
from logging import Logger
from typing import List
from transport import Subscriber, RPCServer

class MainApp(App):

    def __init__(self, logger: Logger, router: Flask, subscribers: List[Subscriber], rpc_servers: List[RPCServer], http_port: int, rmq_connection_list: List[pika.BlockingConnection]):
        self._http_port: int = http_port
        self._liveness: bool = False
        self._readiness: bool = False
        self._logger: Logger = logger
        self._router: Flask = router
        self._subscribers: List[Subscriber] = subscribers
        self._rpc_servers: List[RPCServer] = rpc_servers
        self._rmq_connection_list: List[pika.BlockingConnection] = rmq_connection_list

    def liveness(self) -> bool:
        return self._liveness

    def readiness(self) -> bool:
        return self._readiness

    def set_liveness(self, liveness: bool) -> None:
        self._liveness = liveness

    def set_readiness(self, readiness: bool) -> None:
        self._readiness = readiness

    def setup(self, components: List[Comp]) -> None:
        for component in components:
            component.setup()

    def run(self) -> None:
        try:
            with self._router.app_context():
                try:
                    self._serve_and_subscribe()
                    self._set_liveness_and_readiness(True)
                except Exception as e:
                    self._set_liveness_and_readiness(False)
                    raise e
            self._link_rmq_status_to_app(self._rmq_connection_list)
            self._router.run("0.0.0.0", self._http_port)
        except Exception as e:
            self._set_liveness_and_readiness(False)
            self._logger.error(e)
    
    def _link_rmq_status_to_app(self, rmq_connection_list: List[pika.BlockingConnection]):
        def check_closed():
            is_ok = True
            while is_ok:
                for rmq_connection in rmq_connection_list:
                    if rmq_connection.is_closed:
                        self._set_liveness_and_readiness(False)
                        is_ok = False
                        break
        thread = threading.Thread(target=check_closed)
        thread.start()

    def _set_liveness_and_readiness(self, liveness_and_readiness: bool) -> None:
        self.set_liveness(liveness_and_readiness)
        self.set_readiness(liveness_and_readiness)

    def _serve_and_subscribe(self) -> None:
        for subscriber in self._subscribers:
            subscriber.subscribe()
        for rpc_server in self._rpc_servers:
            rpc_server.serve()